app/controllers/api/base_controller/renderer.rb

```diff
@@ -540,6 +546,47 @@ def render_options(resource, data = {})
         klass = collection_class(resource)
         render :json => OptionsSerializer.new(klass, data).serialize
       end
+
+      def virtual_attributes_for(klass)
+        validate_attr_selection(klass).last.reject { |vattr| vattr == "href_slug" }
+      end
+
+      def determine_include_for_find(klass)
+        return nil unless klass.respond_to?(:reflect_on_association)
+
+        type             = @req.subject
+        include_for_find = []
+
+        virtual_attributes_for(klass).each do |vattr|
+          attr_name, attr_base = split_virtual_attribute(vattr)
+          next if attr_base.blank?
+          next if attr_base.include?(".") # FIXME: Allow nested relations
+          next if virtual_attribute_accessor(type, attr_name)
+          next if Rbac::Filterer::CLASSES_THAT_PARTICIPATE_IN_RBAC.include?(attr_base)
+
+          include_for_find << attr_base
+        end
+
+        include_for_find.empty? ? nil : include_for_find
+      end
+
+      def determine_extra_cols(klass)
+        return nil unless klass.respond_to?(:reflect_on_association)
+
+        type       = @req.subject
+        extra_cols = []
+
+        virtual_attributes_for(klass).each do |vattr|
+          attr_name, attr_base = split_virtual_attribute(vattr)
+          next if attr_base.present?
+          next if virtual_attribute_accessor(type, attr_name)
+          next unless klass.attribute_supported_by_sql?(attr_name)
+
+          extra_cols << attr_name.to_sym
+        end
+
+        extra_cols.empty? ? nil : extra_cols
```

Comment:
Both of these methods look similar.  Is this common behavior something worthy of a named method?

Reply:
I might be able to push some of this shared logic into `virtual_attributes_for`... let me test some things out with that.

Reply:
Well, I'm really wondering what is this common behavior called? It might make sense to name it since even I can see similarities 🤣   So, yeah, if there's parts that can be named as actual things, it might make sense to organize them by the behavior they're doing.  If not, that's fine also.  I'm just commenting on what appeared to be a "thing" that could want a name.

Reply:
> Well, I'm really wondering what is this common behavior is called?

Well, unfortunately a lot of this is dissecting a bunch of already existing methods to pull out the piece(s) that are relevant to create these new ones.  The problem and why it isn't easy to just share these method fragments else where is the area that was pulled from (the `*_to_jbuilder` methods) fetch/expand the attributes for the result as it is generated, and so breaking down the attributes is done slowly in phases for each subcollection.

Where in the case here, we are trying determine if there are any relations to these methods up front so we don't have the N+1s.  The `fetch_direct_virtual_attribute` and `fetch_indirect_virtual_attribute` are where a lot of these came from, but there are some other places that I can't think of off the top of my head as well.

Reply:
@jrafanie I took a stab at refactoring this to DRY things up:

https://github.com/ManageIQ/manageiq-api/pull/887/commits/934ee3e13321f72e002d134e54c56dfbaad6a307

Let me know what you think.